sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedColumnReader.java

```diff
@@ -122,7 +123,24 @@ public VectorizedColumnReader(
     DictionaryPage dictionaryPage = pageReader.readDictionaryPage();
     if (dictionaryPage != null) {
       try {
-        this.dictionary = dictionaryPage.getEncoding().initDictionary(descriptor, dictionaryPage);
+        PrimitiveType primitiveType = descriptor.getPrimitiveType();
+        if (primitiveType.getOriginalType() == OriginalType.DECIMAL &&
+            primitiveType.getDecimalMetadata().getPrecision() <= Decimal.MAX_INT_DIGITS() &&
+            primitiveType.getPrimitiveTypeName() == PrimitiveType.PrimitiveTypeName.INT64) {
+          // We need to make sure that we initialize the right type for the dictionary otherwise
+          // Encoding#initDictionary will initialize it to PrimitiveTypeName.INT64 and
+          // WritableColumnVector will throw an exception when trying to decode to an Int when the
+          // dictionary is in fact initialized as Long
```

Comment:
Hm .. why should we manually modify the column descriptor from Parquet though? Does the regular non-vectorized Parquet reader do this too?

Reply:
I don't know for sure but 3rd party only base the output on parquet schema of the file. 
The reason why we have to do this is because we are initializing the `WritableColumnVector` based on the sparkSchema and thus initializes it based on the precision and anything less than 10 precision is initialized to be backed by an Int even if the parquet schema has it stored as a long.

Reply:
Can we add a test when vectorized reader (`spark.sql.parquet.enableVectorizedReader`) is off too?